### DISCUSSION TITLE
Tips and tricks

### DISCUSSION THREAD

**Discussion by GreatOwle** [d21spike/plugin.video.sling#33] [Q&A]
I am curious about how your systems are configured for add-on development. Would this be a good place to ask questions or should I take this elsewhere? I could start first to get the ball rolling if it is.

**Comment by d21spike** [comment]
This being python and pretty platform independent, there isn't much to setup. More-so just having tools you like to use.

**Editors**
- VSCode (my go-to)
- PyCharm (very good git integration)

Usually I end up checking out the repo into the Kodi addons directory, this way I don't have to reinstall the addon every time but make changes on the fly without even restarting Kodi. Unfortunately, this only lets you develop for one build of Kodi.

To develop for Kodi 18, I utilize a portable version. This way I'm able to keep Matrix installed as default.

For built-in Kodi functions, consulting the documentation usually will get you the functions available to you.
[Kodi Documentation](https://codedocs.xyz/xbmc/xbmc/group__python.html)

If you want code-completion for Kodi libraries, then checkout Kodistubs
[Kodistubs](https://romanvm.github.io/Kodistubs/)

Usually have the `Kodi.log` file open in my text editor, but the browser is also an option. I prefer the text editor since I can pick a syntax and have highlighting. This makes it a bit easier to read. There are addons you can install in Kodi which host a small webserver and you can view the log file real-time. This is especially useful for devices other than your development machine, such as android tv boxes and the like.

Lastly, log .... log ... log. If something is not working, you can print out anything and everything. It's easy to go back and clean up logging.

**Reply by GreatOwle** [reply]
Thank you d21spike. This is exactly the kind of info I was hoping to get. What OS platform(s) are you using?

**Comment by GreatOwle** [comment]
I'm on Ubuntu 20.10.

My tools are:
- PyCharm Community - My editor but mostly using it right now to browse code.
- tree - CLI to get a gross overview of how code is structured.
- meld - File and folder diff utility. Using this mainly to diff (a) d21spike's repository with my own, and (b) KODI inputstream.adaptive with the upstream inputstream.adaptive.
- GitKraken - Great for visualizing branches and commits. Has a built-in diff utility that makes it easy to review changes.

My tricks, such as they are:
- Addon repositories are cloned to ~/kodi-addons. Soft links are added to the ~/.kodi/addons folder to reference any that I want to use. Any plugin I don't want kodi to see can be hidden by renaming its soft link to start with a '.'.
- Soft link for ~/kodi/temp/kodi.log added to desktop for easy access.
- meldcmp files (these are saved settings for any single file or folder comparison) added to desktop for easy access to the two folder comparisons that I am interested in.

More questions:
- Do you ever build Kodi from source?
- What repositories do you follow (other than kodi and inputstream.adaptive)?
- Still looking for a good standalone visual search and replace utility. Any recommendations?
- What about debugging? I was able to use web_pdb to step through python code. Is there a way to use VSCode or PyCharm instead?
